src/content/en/fundamentals/engage-and-retain/push-notifications/common-notification-patterns.md

```diff
@@ -0,0 +1,375 @@
+
+There is also a `notificationclose` event that is called if the user dismisses one of your
+notifications (i.e. rather than click the notification, the user clicks the cross or swipes the
+notification away).
+
+This event is normally used for analytics to track user engagement with notifications.
+
+    self.addEventListener('notificationclose', function(event) {
+      const dismissedNotification = event.notification;
+
+      const promiseChain = notificationCloseAnalytics();
+      event.waitUntil(promiseChain);
+    });
+
+## Adding Data to a Notification
+
+When a push message is received it's common to have data that isn't useful the user has clicked
+the notification. For example, the URL that should be opened if the user clicks on a
+notification. The data can be passed in with the push event but it's not going to be used until
+the `notificationclick` event.
+
+The easiest way to do this is to add data to a notification in the options for
+`showNotification()` like this:
+
+        const options = {
+          body: 'This notification has data attached to it that is printed ' +
+            'to the console when it\'s clicked.',
+          tag: 'data-notification',
+          data: {
+            time: new Date(Date.now()).toString(),
+            message: 'Hello, World!'
+          }
+        };
+        registration.showNotification('Notification with Data', options);
+
+Inside a click handler the data can be accessed with `event.notification.data`.
+
+      const notificationData = event.notification.data;
+      console.log('');
+      console.log('The data notification had the following parameters:');
+      Object.keys(notificationData).forEach((key) => {
+        console.log(`  ${key}: ${notificationData[key]}`);
+      });
+      console.log('');
+
+## Open a Window
+
+One of the most common actions performance when a user clicks on a notification is to open a
+window / tab to a specific URL. We can do this with the
+[clients.openWindow()](https://developer.mozilla.org/en-US/docs/Web/API/Clients/openWindow)
+API.
+
+In our `notificationclick` event we'd run some kind like this:
+
+      const examplePage = '/demos/notification-examples/example-page.html';
+      const promiseChain = clients.openWindow(examplePage);
+      event.waitUntil(promiseChain);
+
+In the next section we'll look at how to check if the page we want to direct the user to is
+already open or not. This way we can focus on the open tab rather than constantly opening new
+tabs.
+
+## Focus an Existing Window
+
+When it's possible, we should focus a window rather than open a new window every time the user
+clicks a notification.
+
+Before we look at how to achieve this, it's worth highlighting that this is **only possible for
+pages on your origin**. This is because we can only see what pages are open that belong to our
+site, this prevents developers from being able to see all the sites their users are viewing.
```

Comment:
'This prevents developers' is a new sentence. Also how about being more direct. 'This prevents sites from spying on their users' browsing.'